resolwe/permissions/shortcuts.py

```diff
@@ -23,48 +61,138 @@ def get_user_group_perms(user_or_group, obj):
+        group_filters = {'pk': group.pk}
     if group_model.objects.is_generic():
         group_filters.update({
-            '%s__content_type' % group_rel_name: ctype,
-            '%s__object_pk' % group_rel_name: obj.pk,
+            '{}__content_type'.format(group_rel_name): ctype,
+            '{}__object_pk'.format(group_rel_name): obj.pk,
         })
     else:
-        group_filters['%s__content_object' % group_rel_name] = obj
-
-    user_perms = []
-    group_perms = []
-
-    if user and user.is_superuser:
-        user_perms = list(chain(*Permission.objects
-                                .filter(content_type=ctype)
-                                .values_list("codename")))
-    elif user:
-        model = get_user_obj_perms_model(obj)
-        related_name = model.permission.field.related_query_name()
-        user_filters = {'%s__user' % related_name: user}
-        if model.objects.is_generic():
-            user_filters.update({
-                '%s__content_type' % related_name: ctype,
-                '%s__object_pk' % related_name: obj.pk,
-            })
-        else:
-            user_filters['%s__content_object' % related_name] = obj
+        group_filters['{}__content_object'.format(group_rel_name)] = obj
+
+    user_perms, group_perms = [], []
+
+    if user:
         perms_qs = Permission.objects.filter(content_type=ctype)
-        user_perms_qs = perms_qs.filter(**user_filters)
-        user_perms = user_perms_qs.values_list("codename", flat=True)
-        group_perms_qs = perms_qs.filter(**group_filters)
-        group_perms = group_perms_qs.values_list("codename", flat=True)
-    else:
-        user_perms = list(set(chain(*Permission.objects
-                                    .filter(content_type=ctype)
-                                    .filter(**group_filters)
-                                    .values_list("codename"))))
+        if user.is_superuser:
+            user_perms = list(chain(perms_qs.values_list("codename", flat=True)))
+        else:
+            model = get_user_obj_perms_model(obj)
+            related_name = model.permission.field.related_query_name()
+            user_filters = {'{}__user'.format(related_name): user}
+            if model.objects.is_generic():
+                user_filters.update({
+                    '{}__content_type'.format(related_name): ctype,
+                    '{}__object_pk'.format(related_name): obj.pk,
+                })
+            else:
+                user_filters['{}__content_object'.format(related_name)] = obj
+
+            user_perms_qs = perms_qs.filter(**user_filters)
+            user_perms = list(chain(user_perms_qs.values_list("codename", flat=True)))
+
+    group_perms_qs = Group.objects.filter(**group_filters)
+    group_perms = list(chain(group_perms_qs.values_list(
+        "pk", "name", "{}__permission__codename".format(group_rel_name))))
+
+    # TODO: Optimize by writing custom Django's aggregate expresion
```

Comment:
What exactly should be optimized here? You just need to ensure that all the returned items are ordered by `(pk, name)` and then you can group in Python in a single pass, which is optimal. I am not sure that generating an array in the database would be much difference unless there are really a lot of rows returned (in which case the overhead of repeating pk and name would be noticeable).